### ISSUE TITLE
Nullable flow annotations do not affect type arguments' inference in the annotated call

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#33105]
**Version Used**: 

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 2.1
VisualStudio.16.Preview/16.0.0-pre.2.1+28529.54
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
using System.Runtime.CompilerServices;
class C
{
  void M1(string? s1)
  {
    s1.NotNull();
    Generic(s1).ToString(); // ok, .NotNull() ensures s1 is not null, Generic.T is inferred to be a non-nullable string
  }

  void M2(string? s1)
  {
    s1.NotNull().ToString(); // false warning - s1.NotNull() should be inspected after applying EnsuresNotNull annotation knowledge
                             // inferring NotNull.T to be not nullable
  }

  void M3(string? s1)
  {
    Ext.NotNull(s1).ToString(); // false warning - NotNull(s1) should be inspected after applying EnsuresNotNull annotation knowledge
                                // inferring NotNull.T to be not nullable
  }

  extern T Generic<T>(T t);
}

static class Ext
{
  public static extern T NotNull<T>([EnsuresNotNull] this T t);
}

namespace System.Runtime.CompilerServices
{
  public class EnsuresNotNullAttribute : System.Attribute { }
}
```

**Expected Behavior**:
No warnings.
Roslyn should infer generic parameters to be a non-nullable T for all the examples above. Since the `NotNull` methods ensures that `s1` is not null the invocation should be reinterpreted with the knowledge that `s1` is a non-nullable string and infer its type argument to be non-nullable string thus allowing derferencing its result.

**Actual Behavior**:
`M2` and `M3` both have the following warning: `CS8602 Possible dereference of a null reference.`.

**Comment by jcouv** [comment]
Here are the LINQ notes specifying the behavior of various attributes. Some introduce warnings (on inbound values) and others modify the null-state of outbound values.
They do not affect type inference.
https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md
